date_range/models/date_range_type.py

```diff
@@ -0,0 +1,44 @@
+# Copyright 2016 ACSONE SA/NV (<http://acsone.eu>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class DateRangeType(models.Model):
+    _name = "date.range.type"
+    _description = "Date Range Type"
+
+    @api.model
+    def _default_company(self):
+        return self.env['res.company']._company_default_get('date.range')
+
+    name = fields.Char(required=True, translate=True)
+    allow_overlap = fields.Boolean(
+        help="If sets date range of same type must not overlap.",
+        default=False)
+    active = fields.Boolean(
+        help="The active field allows you to hide the date range without "
+        "removing it.", default=True)
+    company_id = fields.Many2one(
+        comodel_name='res.company', string='Company', index=1,
+        default=_default_company)
+    date_range_ids = fields.One2many('date.range', 'type_id', string='Ranges')
+
+    _sql_constraints = [
+        ('date_range_type_uniq', 'unique (name,company_id)',
+         'A date range type must be unique per company !')]
+
+    @api.constrains('company_id')
+    def _check_company_id(self):
+        if not self.env.context.get('bypass_company_validation', False):
+            for rec in self.sudo():
+                if not rec.company_id:
+                    continue
+                if bool(self.date_range_ids.filtered(
```

Comment:
I think you don't need the cast to `bool`.
Check on `rec` instead of `self`.

Reply:
@SimoRubi I think you misunderstand the code..... 
We check if a date_range with another company `lambda r: r.company_id and  r.company_id != rec.company_id` already exists into the list of date ranges (self.date_range_ids) with the same type.

Reply:
Thanks for the clarification @lmignon, but if `self` is a recordset, then wouldn't `self.date_range_ids` throw an Exception?

Reply:
@SimoRubi You're right, the logic here is not right.... may be something like
```python
for rec in self.sudo():
    company_ids = filtered(None, rec.date_range_ids.mapped('company_id'))
    if company_ids and company_ids != rec.company_id:
        raise ....
```
to be tested

Reply:
But replacing self by rec is even more easy and readable ;-)

Reply:
With latest commit, the self is replaced by rec. Is that enough? Do you all agree?